### PR TITLE
[libc++][test] Workaround for atomic tests linker errors on Linux systems without libatomic.so

### PR DESCRIPTION
**Pull request by huixie90**
- [libc++][test] fix atomic float test CI failure
- format
- [libc++][test] atomic tests add latomic workaround

### CODE REVIEW

File: libcxx/utils/libcxx/test/features.py
Comment (ldionne) [inline libcxx/utils/libcxx/test/features.py:289]: This part of the diff looks like a merge conflict to me.

**Review by ldionne — APPROVED** [review]
I see this in https://buildkite.com/llvm-project/libcxx-ci/builds/31901#018c15d9-3491-4b6f-b2d5-92ce78d2c7a9

```
Unexpectedly Passed Tests (17):
  libcxx/atomics/atomics.types.generic/atomics.types.float/lockfree.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/assign.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/compare_exchange_strong.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/compare_exchange_weak.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/ctor.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/exchange.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/fetch_add.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/fetch_sub.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/load.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/lockfree.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/notify_all.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/notify_one.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/operator.float.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/operator.minus_equals.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/operator.plus_equals.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/store.pass.cpp
  std/atomics/atomics.types.generic/atomics.types.float/wait.pass.cpp
```

I think it means that we're failing to detect `non-lockfree-atomics` on Windows, yet the code apparently seems to work. To unblock this patch and restore some CI stability, we could mark these tests as `UNSUPPORTED: !non-lockfree-atomics` and investigate the Windows problems separately. I suspect it would be really easy for @mstorsjo to help us out here.

**Review by ldionne — APPROVED** [review]
LGTM assuming this passes CI (and I think it should).

**Comment by zeroomega** [comment]
Is this patch ready to land? I run through tests on our bots and can confirm it clears the atomic test failures. It would be great to land this patch timely.

**Comment by mstorsjo** [comment]
> I think it means that we're failing to detect `non-lockfree-atomics` on Windows, yet the code apparently seems to work. To unblock this patch and restore some CI stability, we could mark these tests as `UNSUPPORTED: !non-lockfree-atomics` and investigate the Windows problems separately. I suspect it would be really easy for @mstorsjo to help us out here.

I think the reason for these cases is pretty simple; when compiling a piece of code that uses atomics, there's nothing that says that the compiler _MUST_ generate a call to `__atomic_load` or similar. The compiler is free to decide whether it is better to generate such code inline or generate a function call to a helper function.

So for such environments where the atomics helper functions are unavailable, some tests may succeed and some may fail, depending on what code generation decisions the compiler makes - thus, one can't really mark them `XFAIL`.

**Comment by ldionne** [comment]
> 
> So for such environments where the atomics helper functions are unavailable, some tests may succeed and some may fail, depending on what code generation decisions the compiler makes - thus, one can't really mark them `XFAIL`.

We are pretty seriously starting to need compiler-rt on Windows though. This means that atomics on Windows will either work or fail to link depending on how much the compiler decides to inline. And this is also blocking us from refactoring our implementation of `std::complex` to use vastly superior builtins. What would be the path for this to happen?


> Is this patch ready to land? I run through tests on our bots and can confirm it clears the atomic test failures. It would be great to land this patch timely.

Yes, I am merging it now. We were just waiting for the CI to complete over night.

**Comment by mstorsjo** [comment]
> > So for such environments where the atomics helper functions are unavailable, some tests may succeed and some may fail, depending on what code generation decisions the compiler makes - thus, one can't really mark them `XFAIL`.
> 
> We are pretty seriously starting to need compiler-rt on Windows though.

This is a more nuanced and complex matter than that, though.

The issue that has been discussed multiple times before, is that MSVC style environments don't link in compiler-rt at all - but it should, ideally. There are plans on how to fix that, there was some progress on it earlier, but there still are a couple steps missing for that. (I'm not personally actively working on that, but I do comment on the plans for how they intend to do it.)

For MinGW environments, which we also test, we already do have compiler-rt builtins. However, by default, the compiler-rt builtins library doesn't include atomics: https://github.com/llvm/llvm-project/blob/llvmorg-17.0.6/compiler-rt/lib/builtins/CMakeLists.txt#L221-L223

> skip the atomic builtin (these should normally be provided by a shared library)

When on a platform where there's a separate shared atomic library, that is supposed to be used, but here, I guess one should either enable this (i.e. disable `COMPILER_RT_EXCLUDE_ATOMIC_BUILTIN`), or enable a different option `COMPILER_RT_BUILD_STANDALONE_LIBATOMIC`, https://github.com/llvm/llvm-project/blob/llvmorg-17.0.6/compiler-rt/lib/builtins/CMakeLists.txt#L857-L859, which builds a separate `libclang_rt.atomic`. I'm not sure if Clang links that one in automatically if it exists or what the story around it is.

For the MinGW builds, I could try to look into whether we should do either of these, until the LLVM 18 releases (we're past all planned 17.x releases for now), so we could have this in libcxx once 18 is out.

**Pull request merged**
